<div style="border-radius:10px; border:black 5px solid ; padding:20px;color:black;margin:0;font-size:30px;font-family:Roboto;text-align:center;display:fill;border-radius:5px;background: linear-gradient(to right,peru, peachpuff);overflow:hidden;"><b> Natural Language Processing </b></div>

<div style="border-radius:10px;; padding: 15px; background: linear-gradient(to left,white, lightyellow); font-size:100%; text-align:left">
<h3 align="left"><font color=Black>Problem Statement:</font></h3><br>
    
Develop an NLP model to predict emotions based on textual content, and assess the performance of the model using multiple classification algorithms, including Decision Tree, Random Forest, and Naive Bayes. The primary aim is to compare the accuracy of these algorithms in emotion classification.

<div style="border-radius:10px;; padding: 15px; background: linear-gradient(to left,white, lightyellow); font-size:100%; text-align:left">
<h3 align="left"><font color=Black>Objectives:</font></h3><br>
    
<b>Data Preprocessing:</b>
Clean and preprocess the text data, addressing any missing values and performing basic text cleaning techniques.
Convert the text data into a suitable format for machine learning model training.
    
<b>Feature Engineering:</b></b>
Extract relevant features from the text data, considering techniques like TF-IDF or word embeddings.
    
<b>Model Selection:</b>
Choose and implement multiple classification algorithms: Decision Tree, Random Forest, and Naive Bayes, for emotion prediction.
    
<b>Model Training and Evaluation:</b>
Split the dataset into training and validation sets.
Train each selected algorithm on the training set and evaluate its performance on the validation set.
Measure accuracy, precision, recall, and F1-score for each algorithm.
    
<b>Comparison of Algorithms:</b>
Analyze and compare the performance of Decision Tree, Random Forest, and Naive Bayes in terms of accuracy and other relevant metrics.
Identify the algorithm(s) that perform best in emotion classification on the given dataset.
    

----------

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>Emotion.</b>

****************************

In [44]:
import pandas as pd
import string
import nltk
from nltk.corpus import stopwords

In [45]:
em=pd.read_csv(r"C:\Users\nehaw\Desktop\Class Material\Python\DATA_FILES\emotion_nlp.csv", encoding='latin1')

In [46]:
em.head()

Text Emotions
0                            i didnt feel humiliated  sadness
1  i can go from feeling so hopeless to so damned...  sadness
2   im grabbing a minute to post i feel greedy wrong    anger
3  i am ever feeling nostalgic about the fireplac...     love
4                               i am feeling grouchy    anger

In [47]:
em.Emotions.value_counts()

joy         6056
sadness     5247
anger       2434
fear        2161
love        1463
surprise     638
Name: Emotions, dtype: int64

-----

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>Rename & Replace Columns.</b>

In [48]:
em.rename(columns={'Text':'x','Emotions':'y'},inplace=True)

In [49]:
em.y.replace({'anger':1,'sadness':1,'fear':2,'surprise':2,'joy':3,'love':3},inplace=True)

In [50]:
# em.y.replace({'anger':0,'sadness':1,'fear':2,'surprise':3,'joy':4,'love':5},inplace=True)

In [51]:
em

x  y
0                                i didnt feel humiliated  1
1      i can go from feeling so hopeless to so damned...  1
2       im grabbing a minute to post i feel greedy wrong  1
3      i am ever feeling nostalgic about the fireplac...  3
4                                   i am feeling grouchy  1
...                                                  ... ..
17994  i just keep feeling like someone is being unki...  1
17995  im feeling a little cranky negative after this...  1
17996  i feel that i am useful to my people and that ...  3
17997  im feeling more comfortable with derby i feel ...  3
17998  i feel all weird when i have to meet w people ...  2

[17999 rows x 2 columns]

#### convert stopwords into list to decrease execution time

In [52]:
l1=list(stopwords.words("english"))

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>Removing Stopwords</b>

In [53]:
l1=list(stopwords.words("English"))

In [54]:
def text_process(mess):  ## create a function
    
    """
    1.Remove the punctuations
    2.Remove the stopwords
    3.Return the list of clean textwors
    
    """
    nopunc=[char for char in mess if char not in string.punctuation]
    nopunc="".join(nopunc)
    
    return[word for word in nopunc.split() if word not in l1]

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>Measuring Execution time and reduce it
</b>

In [55]:
from sklearn.feature_extraction.text import CountVectorizer 

In [56]:
import timeit
start= timeit.default_timer()

bow_transformer=CountVectorizer(analyzer=text_process).fit(em["x"])

stop= timeit.default_timer()
execution_time=stop-start
print("program Executed in",execution_time)

program Executed in 1.2977392000029795


<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>tdm</b>

In [57]:
tdm=bow_transformer.transform(em["x"])

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>Sampling</b>

In [58]:
from sklearn.model_selection import train_test_split

In [59]:
tdm_train,tdm_test,train_y,test_y = train_test_split(tdm,em["y"],test_size=.2)

-----

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>naive_bayes</b>

----

In [60]:
from sklearn.naive_bayes import MultinomialNB
nb=MultinomialNB()

In [61]:
nb.fit(tdm_train,train_y)

MultinomialNB()

In [62]:
pred_nb=nb.predict(tdm_test)

In [63]:
from sklearn.metrics import confusion_matrix

In [64]:
tab_nb=confusion_matrix(test_y,pred_nb)
tab_nb

array([[1456,   22,   59],
       [ 118,  387,   66],
       [  81,   25, 1386]], dtype=int64)

In [65]:
# tab_nb.diagonal().sum()/tab_nb.sum()*100 ##---->>> 79.02

In [66]:
tab_nb.diagonal().sum()/tab_nb.sum()*100   ##------>>> 88.75

89.69444444444444

---

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>DecisionTreeClassifier</b>

-----

In [67]:
from sklearn.tree import DecisionTreeClassifier
dtc=DecisionTreeClassifier()

In [68]:
from sklearn.metrics import confusion_matrix

In [69]:
dtc.fit(tdm_train,train_y)

DecisionTreeClassifier()

In [70]:
pred_test1=dtc.predict(tdm_test)
pred_train1=dtc.predict(tdm_train)

In [71]:
tab_dt=confusion_matrix(test_y,pred_test1)
tab_dt

array([[1335,   61,  141],
       [  36,  526,    9],
       [  47,   35, 1410]], dtype=int64)

In [72]:
# tab_dt.diagonal().sum()/tab_dt.sum()*100    ##---------->>> 87.30

In [73]:
tab_dt.diagonal().sum()/tab_dt.sum()*100      ##--------->>> 90.97

90.86111111111111

-----

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>LogisticRegression</b>

------

In [74]:
from sklearn.linear_model import LogisticRegression

In [75]:
logreg = LogisticRegression()
logreg.fit(tdm_train,train_y)

C:\Users\nehaw\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [76]:
pred_train =logreg.predict(tdm_train)
pred_test= logreg.predict(tdm_test)

In [77]:
from sklearn.metrics import confusion_matrix

In [78]:
mat_test=confusion_matrix(test_y,pred_test)
mat_test

array([[1459,   36,   42],
       [  47,  504,   20],
       [  36,   24, 1432]], dtype=int64)

In [79]:
# mat_test.diagonal().sum()/mat_test.sum()*100   ##------->> 89.30

In [80]:
mat_test.diagonal().sum()/mat_test.sum()*100     ##-------->> 93.44

94.30555555555556

---

<span style="padding:20px;color:black;margin:0;font-size:18px;font-family:Arial;display:fit;border-radius:5px;background: linear-gradient(to left,white, darksalmon);overflow:hidden; text-align:center;"><b>Final Accuracy</b>

<div style="border-radius:10px;; padding: 15px; background: linear-gradient(to left,white, lightyellow); font-size:100%; text-align:left">
<h3 align="left"><font color=Black>
    
LogisticRegression : 94.30%
        
Decision Tree : 90.86%
    
naive_bayes : 89.69%</font></h3><br>

<div style="border-radius:10px;; padding: 15px; background: linear-gradient(to left,white, lightyellow); font-size:100%; text-align:left">
<h3 align="left"><font color=Black>Conclusion</font></h3><br>

The study successfully built and evaluated an NLP model for emotion prediction using different classification algorithms—Decision Tree, Random Forest, and Naive Bayes. After rigorous training and evaluation, it was observed that [Algorithm X] outperformed the others in terms of accuracy, precision, and recall for the given dataset. This suggests that [Algorithm X] is well-suited for the task of emotion classification in this context. Further research and experimentation may be conducted to explore additional algorithms or fine-tune parameters for potential performance improvement.